src/libos/src/time/timer_file.rs

```diff
@@ -0,0 +1,394 @@
+            let inner_c = self.inner.clone();
+
+            async move {
+                loop {
+                    let waiter = Waiter::new();
+                    // The initial timeout is the it_value.
+                    let res = waiter.wait_timeout(Some(&mut timeout)).await;
+                    match res {
+                        Err(e) => {
+                            if e.errno() == EINTR {
+                                // Error or Stopped by tirq
+                                break;
+                            } else if one_shot == true {
+                                // One-shot timer, no need loop
+                                trace!("timerfd one-shot triggerred");
+                                pollee.add_events(Events::IN);
+                                let mut inner = inner_c.lock().unwrap();
+                                inner.exp_cnt += 1;
+                                inner.status = TimerFdStatus::EXPIRE;
+                                inner.task_handle = None;
+                                break;
+                            } else {
+                                trace!("timerfd timer expired");
+                                // If it is not one-shot, the following timeout changes to interval.
+                                timeout = interval;
+                                pollee.add_events(Events::IN);
+                                let mut inner = inner_c.lock().unwrap();
+                                inner.exp_cnt += 1;
+                            }
+                        }
+                        _ => {
+                            panic!("impossible as there is no waker or timeout");
+                        }
+                    }
+                }
+
+                trace!("Timerfd poll task end");
+            }
+        })
+        .priority(async_rt::sched::SchedPriority::Low)
+        .spawn();
+
+        inner.task_handle = Some(join_handle);
+
+        Ok(())
+    }
+
+    async fn wait_for_expire(&self) -> Result<usize> {
+        let flags = self.flags.load(Ordering::Relaxed);
+        let mut count: usize = 0;
+
+        let inner_c = self.inner.clone();
+        let mut inner = inner_c.lock().unwrap();
+        if inner.status == TimerFdStatus::STOP {
+            return_errno!(EAGAIN, "try again");
+        } else if inner.status == TimerFdStatus::EXPIRE {
+            count = inner.exp_cnt - inner.last_exp_cnt;
+            return Ok(count);
+        }
+
+        drop(inner);
+
+        loop {
+            let mut inner = inner_c.lock().unwrap();
+            count = inner.exp_cnt - inner.last_exp_cnt;
+
+            // Do not always hold it in the loop
+            drop(inner);
+
+            if count > 0 {
+                break;
+            } else if flags.contains(TimerCreationFlags::TFD_NONBLOCK) {
+                return_errno!(EAGAIN, "try again");
+            } else {
+                // Block wait, do yield
+                yield_().await;
```

Comment:
It is not an efficient way. We should make sure the timer is triggered before check the exp_cnt.